src/Control/Distributed/Spark/Iterator.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Control.Distributed.Spark.Iterator
+  ( Iterator(..)
+  , JIterator
+  , emptyIterator
+  , foldIterator
+  , singletonIterator
+  ) where
+
+import Control.Distributed.Closure.TH
+import Data.Int
+import Data.IORef
+import Foreign.JNI
+import Foreign.Ptr
+import Foreign.StablePtr
+import Language.Java
+import System.IO (fixIO)
+
+
+-- | This is a haskell representation for iterators.
+newtype Iterator a = Iterator (IO (Maybe (a, Iterator a)))
+type JIterator a = 'Iface "java.util.Iterator" <> '[a]
+type instance Interp (Iterator a) = JIterator (Interp (Uncurry a))
+
+type JHaskellIterator = J ('Class "io.tweag.sparkle.HaskellIterator")
```

Comment:
We don't define such type synonyms anywhere else in the sparkle API. If you made `Iterator` be a newtype wrapper around `J ('Class "io.tweag.sparkle.HaskellIterator")`, then you could dispense with it. Users could then use `classOf` to get the class name if they really need it, like they do for all other exported types in sparkle.

Indeed `Iterator` should be the reflection of a Haskell type like `Conduit`, rather than an actual Haskell iterator.